jetson/nav/recovery.cpp

```diff
@@ -0,0 +1,206 @@
+#include "recovery.hpp"
+#include "utilities.hpp"
+
+Recovery::Recovery(std::chrono::duration<int> offset, double dispThresh,
+                   std::chrono::duration<int> maxTurnTime) : offsetAmount(offset), dispThresh(dispThresh), maxTurnTime(maxTurnTime) {
+
+}// Recovery()
+
+bool Recovery::isStuck() {
+    // check displacement and or turn time to detect if rover is stuck
+    // if turning and time_turning > maxTurnTime: true
+    // if not turning and displacement < threshold: return true
+    // else return false
+
+    if (wasTurning) {
+        return getDuration(odoms.front().time, turnStartTime) > maxTurnTime;
+    } else {
+
+        // distance based
+        double distance = estimateDistance(odoms.front().odom, odoms.back().odom);
+
+        // note:  the displacement time stamp to statisfy condition must be >= offsetAmount
+        // but we also clear all odoms > offsetAmount
+        return (!odoms.empty() && getDuration(odoms.front().time, odoms.back().time) >= offsetAmount && distance < dispThresh);
+    }
+
+
+}// isStuck()
+
+void Recovery::update(Odometry currOdom, time_point currTime, bool isTurning) {
+    // updates odoms deque with new data and remove older data
+    // updates turning state (switch between time based and displacement based stuck detection)
+
+    // TODO: add fail safe for overloading deque with data and making program fail
+    //       may be better to just make a circular buffer instead of using deque
+
+    // update turning state
+    // if we were not turning but we are now turning
+    if (!wasTurning && isTurning) {
+        wasTurning = true;
+        turnStartTime = currTime;
+    }
+
+    // if we were turning and are now not turning
+    else if (wasTurning && !isTurning) {
+        //reset displacement (as we were rotating not being displaced)
+        wasTurning = false;
+        odoms.clear();
+    }
+
+    // add latest odom
+    OdomTimePoint newOdom = {currOdom, currTime};
+    odoms.push_front(newOdom);
+
+    // check old odoms for removal (if newer odoms that are within offset time)
+    // check oldest odom (and second to last odom to see if we should remove the oldest odom)
+    while (odoms.size() > 2 && (getDuration(odoms.front().time, odoms[odoms.size() - 2].time)) > offsetAmount) {
+        odoms.pop_back();
+    }
+
+}// update()
+
+
+// get duration between two points in seconds
+// let p1 be the newer of the time points and p2 be the later of the time points
+std::chrono::duration<int> Recovery::getDuration(time_point p1, time_point p2) {
+    auto seconds = std::chrono::duration_cast<std::chrono::seconds>(p1 - p2);
+    return seconds;
+}// getDuration()
+
+// returns the point the rover should be follwing for the recovery path
+RecoveryPoint Recovery::getPointToFollow() {
+    return currentRecoverypoint;
+}// getPointToFollow()
+
+// generate recovery path J turn
+void Recovery::makeRecoveryPath(const std::shared_ptr<Rover>& rover) {
```

Comment:
I'm honestly not a fan of passing smart pointers in function signatures, even by reference. In any modern C++ codebase it should be fairly clear that raw pointer handoffs do not signify change in ownership. Not a huge deal though.

Reply:
or even better, const reference

Reply:
just to clarify, I meant underlying object as in the object that the smart ptr contains.

Reply:
this should be resolved now